File: server/core/media_organizer.py
"""
Media organizer — moves downloaded files into the library.

New unified layout
------------------
Movies:
  {MEDIA_DIR}/{Title} [{tmdb_id}]/{Title} ({Year}).ext

TV / Anime:
  {MEDIA_DIR}/{Title} [{tmdb_id}]/S{NN}E{NN} - {Episode Title}.ext
  Season packs keep original filenames.
"""
from __future__ import annotations

import logging
import re
import shutil
from pathlib import Path
from typing import Optional

from server.config import settings

logger = logging.getLogger(__name__)

# Characters not allowed in Windows filenames (excluding colon, handled separately)
_ILLEGAL_CHARS = re.compile(r'[<>"/\\|?*]')
# Collapse multiple spaces
_MULTI_SPACE = re.compile(r"\s{2,}")

# Common video extensions
VIDEO_EXTENSIONS = {".mkv", ".mp4", ".avi", ".mov", ".wmv", ".flv", ".webm"}


def _sanitize(name: str) -> str:
    """Remove characters illegal on Windows/macOS filesystems."""
    if not name:
        return ""
    # Replace colon with " - "
    name = name.replace(":", " - ")
    # Remove other illegal characters
    name = _ILLEGAL_CHARS.sub("", name)
    # Collapse multiple spaces
    name = _MULTI_SPACE.sub(" ", name)
    # Strip leading/trailing dots and spaces
    name = name.strip(" .")
    return name


def _pick_video_file(directory: Path) -> Optional[Path]:
    """Return the largest video file in *directory* (recursive)."""
    best: Optional[Path] = None
    best_size = 0
    for p in directory.rglob("*"):
        if p.is_file() and p.suffix.lower() in VIDEO_EXTENSIONS:
            size = p.stat().st_size
            if size > best_size:
                best_size = size
                best = p
    return best


class MediaOrganizer:
    def __init__(self):
        pass

    def _destination(self, video_path: Path, media) -> Path:
        """Build the destination path for a video file based on media info."""
        ext = video_path.suffix
        media_dir = Path(settings.MEDIA_DIR)
        title = _sanitize(media.title)
        tmdb_id = media.tmdb_id
        folder_name = f"{title} [{tmdb_id}]"

        if media.type == "movie":
            if media.year:
                file_name = f"{title} ({media.year}){ext}"
            else:
                file_name = f"{title}{ext}"
            return media_dir / folder_name / file_name

        # TV / Anime
        season = media.season if media.season is not None else 1

        if media.episode is not None:
            ep_title = ""
            if hasattr(media, "episode_titles") and media.episode_titles:
                ep_title = media.episode_titles.get(media.episode, "")

            if ep_title:
                file_name = f"S{season:02d}E{media.episode:02d} - {_sanitize(ep_title)}{ext}"
            else:
                file_name = f"S{season:02d}E{media.episode:02d}{ext}"
            return media_dir / folder_name / file_name
        else:
            # Season pack — keep original filename
            file_name = video_path.name
            return media_dir / folder_name / file_name

    def organize(self, source: Path, media) -> Path:
        """
        Move *source* (a file or directory) to the appropriate library location.
        Returns the final file path.
        """
        # Resolve the actual video file if source is a directory
        if source.is_dir():
            video_file = _pick_video_file(source)
            if not video_file:
                raise FileNotFoundError(f"No video file found in {source}")
        else:
            video_file = source

        dest = self._destination(video_file, media)
        dest.parent.mkdir(parents=True, exist_ok=True)

        # Overwrite if exists
        if dest.exists():
            dest.unlink()

        logger.info("Moving %s → %s", video_file, dest)
        shutil.move(str(video_file), str(dest))

        # Clean up empty staging folders
        try:
            if source.is_dir() and source != video_file.parent:
                shutil.rmtree(source, ignore_errors=True)
        except Exception:
            pass

        return dest

File: app/MediaDownloader/MediaDownloader.Server/Core/MediaOrganizer.cs
using System.Text.RegularExpressions;
using MediaDownloader.Server.Clients;
using MediaDownloader.Server.Configuration;

namespace MediaDownloader.Server.Core;

/// <summary>
/// Moves downloaded files into the library in Plex-compatible paths.
/// Mirrors server/core/media_organizer.py.
/// </summary>
public class MediaOrganizer
{
    private static readonly Regex IllegalChars = new(@"[<>:""/\\|?*\x00-\x1f]", RegexOptions.Compiled);
    private static readonly Regex MultiSpace = new(@"\s{2,}", RegexOptions.Compiled);
    private static readonly HashSet<string> VideoExtensions = [".mkv", ".mp4", ".avi", ".m2ts", ".ts", ".mov", ".wmv", ".flv"];

    private readonly ServerSettings _settings;
    private readonly ILogger? _logger;

    public MediaOrganizer(ServerSettings settings, ILogger? logger = null)
    {
        _settings = settings;
        _logger = logger;
    }

    public string Organize(string sourcePath, MediaInfo media)
    {
        var source = new FileInfo(sourcePath);
        string videoFilePath;

        if (Directory.Exists(sourcePath))
        {
            videoFilePath = PickVideoFile(sourcePath) ?? throw new FileNotFoundException($"No video file found in {sourcePath}");
        }
        else
        {
            videoFilePath = sourcePath;
        }

        var dest = GetDestination(videoFilePath, media);
        Directory.CreateDirectory(Path.GetDirectoryName(dest)!);

        _logger?.LogInformation("Moving {Source} → {Dest}", videoFilePath, dest);
        File.Move(videoFilePath, dest, overwrite: false);

        // Clean up empty staging folders
        if (Directory.Exists(sourcePath) && sourcePath != Path.GetDirectoryName(videoFilePath))
        {
            try { Directory.Delete(sourcePath, recursive: true); } catch { }
        }

        return dest;
    }

    private string GetDestination(string videoFile, MediaInfo media)
    {
        var ext = Path.GetExtension(videoFile).ToLowerInvariant();
        if (string.IsNullOrEmpty(ext)) ext = ".mkv";
        var baseDir = GetBaseDir(media);

        if (media.Type == "movie")
        {
            var folderName = Sanitize(media.Year.HasValue ? $"{media.Title} ({media.Year})" : media.Title);
            var fileName = Sanitize(media.Year.HasValue ? $"{media.Title} ({media.Year}){ext}" : $"{media.Title}{ext}");
            return Path.Combine(baseDir, folderName, fileName);
        }

        // TV / Anime
        var showDir = Sanitize(media.Title);
        var seasonNum = media.Season ?? 1;
        var seasonDir = $"Season {seasonNum:D2}";

        string fileName2;
        if (media.Episode.HasValue)
        {
            var epTitle = media.EpisodeTitles.GetValueOrDefault(media.Episode.Value, "");
            var epSuffix = !string.IsNullOrEmpty(epTitle) ? $" - {Sanitize(epTitle)}" : "";
            fileName2 = Sanitize($"{media.Title} - S{seasonNum:D2}E{media.Episode:D2}{epSuffix}{ext}");
        }
        else
        {
            // Season pack — keep original filename
            fileName2 = Sanitize(Path.GetFileNameWithoutExtension(videoFile) + ext);
        }

        return Path.Combine(baseDir, showDir, seasonDir, fileName2);
    }

    private string GetBaseDir(MediaInfo media) => media.Type switch
    {
        "anime" => _settings.AnimeDir,
        "tv" => _settings.TvDir,
        _ => _settings.MoviesDir,
    };

    private static string Sanitize(string name)
    {
        name = IllegalChars.Replace(name, "");
        return MultiSpace.Replace(name, " ").Trim(' ', '.');
    }

    private static string? PickVideoFile(string directory)
    {
        string? best = null;
        long bestSize = 0;

        foreach (var file in Directory.EnumerateFiles(directory, "*", SearchOption.AllDirectories))
        {
            if (!VideoExtensions.Contains(Path.GetExtension(file).ToLowerInvariant()))
                continue;
            var size = new FileInfo(file).Length;
            if (size > bestSize)
            {
                bestSize = size;
                best = file;
            }
        }
        return best;
    }
}
